Merge PSHOrg cAppPool to xWebAppPool 2

zuehlaa opened the pull request:
Second attempt at this.

tysonjhayes commented:
After reviewing this again I've got some suggestions.

* You shouldn't be modifying the system in a unit test, modifications should be done in an integration test setting.  

You noted you've never used Pester before, and that's OK! We all have to learn, god knows I had to learn with this. :grin:  

I'd recommend a couple of modifications to achieve this goal.

1) Add the following function to your xWebAppPool.psm1

````powershell
function Invoke-AppCmd
{
    param
    (
        [string[]] $Arguments,
        [string] $Path = "$env:SystemRoot\system32\inetsrv\appcmd.exe"
    )

    if (Test-Path $Path)
    {
        & $Path $Arguments
    }
    else
    {
        throw "$Path does not exist"
    }
}
````

Then anytime you call appcmd.exe instead of calling the exe directly call the function. So on line 800 the call becomes this:

````powershell
    $AppPool = Invoke-AppCmd -Arguments list,apppool,$Name
````

This allows you to mock Invoke-AppCmd and return whatever you desire.

````powershell
Context 'Test-TargetResource' {
    It 'Passes test when App Pool does exist and has the correct name'{
        $testParams = @{
            Name = "PesterAppPool"
            Ensure = "Present"
        }

        Mock Invoke-AppCmd {return $testParams}

        Test-TargetResource @testParams | Should be $true
    }

It 'returns false if autoStart is different' {

    $testParams = @{
        Name = "PesterAppPool"
        Ensure = "Present"
        AutoStart = 'False'
    }

    $failParams = $testParams
    $failParams['AutoStart'] = 'True'

    Mock Invoke-AppCmd {return $failParams}

    Test-TargetResource @testParams | Should be $false
}
````

Doing this allows you to run the same tests but control the outcome without having to make any modifications to the system.

tysonjhayes commented on DSCResources/MSFT_xWebAppPool/MSFT_xWebAppPool.schema.mof:
```diff
@@ -1,12 +1,58 @@
 
-[ClassVersion("1.0.0.0"), FriendlyName("xWebAppPool")]
+[ClassVersion("1.0.0.1"), FriendlyName("xWebAppPool")]
```
Don't change this please.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,848 @@
+
+Import-Module (Join-Path $here -ChildPath "..\..\DSCResources\$ModuleName\$ModuleName.psm1")
+
+if (! (Get-Module xDSCResourceDesigner))
```
So this isn't needed anymore as we pulled it out into the common tests.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,848 @@
+                Ensure = "Present"
+            }
+
+            It 'Should pass Test-xDscResource Schema Validation' {
```
This isn't needed anymore because we do it in common tests.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,848 @@
+$here = Split-Path -Parent $MyInvocation.MyCommand.Path
```
Ok so from line 1 to line 22 (Describe "Resource...") I would replace all of it with this:

````powershell
# Check if WebServer is Installed
if (@(Get-WindowsOptionalFeature -Online -FeatureName 'IIS-WebServer' `
    | Where-Object -Property State -eq 'Disabled').Count -gt 0)
{
    if ((Get-CimInstance Win32_OperatingSystem).ProductType -eq 1)
    {
        # Desktop OS
        Enable-WindowsOptionalFeature -Online -FeatureName 'IIS-WebServer'
    }
    else
    {
        # Server OS
        Install-WindowsFeature -IncludeAllSubFeature -IncludeManagementTools -Name 'Web-Server'
    }
}

$DSCModuleName  = 'xWebAdministration'

$moduleRoot = "${env:ProgramFiles}\WindowsPowerShell\Modules\$DSCModuleName"

if(-not (Test-Path -Path $moduleRoot))
{
    $null = New-Item -Path $moduleRoot -ItemType Directory
}
else
{
    # Copy the existing folder out to the temp directory to hold until the end of the run
    # Delete the folder to remove the old files.
    $tempLocation = Join-Path -Path $env:Temp -ChildPath $DSCModuleName
    Copy-Item -Path $moduleRoot -Destination $tempLocation -Recurse -Force
    Remove-Item -Path $moduleRoot -Recurse -Force
    $null = New-Item -Path $moduleRoot -ItemType Directory
}

Copy-Item -Path $PSScriptRoot\..\..\* -Destination $moduleRoot -Recurse -Force -Exclude '.git'

if (Get-Module -Name $DSCModuleName -All)
{
    Get-Module -Name $DSCModuleName -All | Remove-Module
}

Import-Module -Name $(Get-Item -Path (Join-Path $moduleRoot -ChildPath "$DSCModuleName.psd1")) -Force

if (($env:PSModulePath).Split(';') -ccontains $pwd.Path)
{
    $script:tempPath = $env:PSModulePath
    $env:PSModulePath = ($env:PSModulePath -split ';' | Where-Object {$_ -ne $pwd.path}) -join ';'
}


$executionPolicy = Get-ExecutionPolicy
if ($executionPolicy -ne 'Unrestricted')
{
    Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Force
    $rollbackExecution = $true
}
````

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,848 @@
+$here = Split-Path -Parent $MyInvocation.MyCommand.Path
+$ModuleName = "MSFT_xWebAppPool"
```
Also if you could please move the file into Tests\Unit please.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,848 @@
+            {
+                Remove-WebAppPool -Name 'PesterAppPool' -ErrorAction Stop
+            }
+        }        
```
Add the following to rollback execution policies and module paths:

````powershell
    if ($rollbackExecution)
    {
        Set-ExecutionPolicy -ExecutionPolicy $executionPolicy -Force
    }

    if ($script:tempPath) {
        $env:PSModulePath = $script:tempPath
    }
````

zuehlaa commented:
I've updated the psm1 file with Invoke-AppCmd and changed all the appcmd.exe calls.  When I run that through my current tests it passes everything.  So next I went to update the tests file and use the mock Invoke-AppCmd.  I'm calling it inside of a InModuleScope MSFT_xWebAppPool but I'm getting this error

```Internal error detected:  Mock for 'Invoke-AppCmd' in module 'MSFT_xWebAppPool' was called, but does not exist in the mock table.
   at line: 734 in C:\Program Files\WindowsPowerShell\Modules\Pester\Functions\Mock.ps1```

Any ideas?  I've looked around but haven't been able to find anything yet.  I had to modify what you suggested a bit to get it running on my box.  This may be what is causing the issue.  I haven't been able to figure out how to run my test like AppVeyor is running it.  

Here is the first part of my test code

```
# Check if WebServer is Installed
if (@(Get-WindowsOptionalFeature -Online -FeatureName 'IIS-WebServer' `
    | Where-Object -Property State -eq 'Disabled').Count -gt 0)
{
    if ((Get-CimInstance Win32_OperatingSystem).ProductType -eq 1)
    {
        # Desktop OS
        Enable-WindowsOptionalFeature -Online -FeatureName 'IIS-WebServer'
    }
    else
    {
        # Server OS
        Install-WindowsFeature -IncludeAllSubFeature -IncludeManagementTools -Name 'Web-Server'
    }
}

$DSCModuleName  = 'xWebAdministration'

$moduleRoot = "${env:ProgramFiles}\WindowsPowerShell\Modules\$DSCModuleName"

if(-not (Test-Path -Path $moduleRoot))
{
    $null = New-Item -Path $moduleRoot -ItemType Directory
}
else
{
    # Copy the existing folder out to the temp directory to hold until the end of the run
    # Delete the folder to remove the old files.
    $tempLocation = Join-Path -Path $env:Temp -ChildPath $DSCModuleName
    Copy-Item -Path $moduleRoot -Destination $tempLocation -Recurse -Force
    Remove-Item -Path $moduleRoot -Recurse -Force
    $null = New-Item -Path $moduleRoot -ItemType Directory
}

Copy-Item -Path $PSScriptRoot\..\..\* -Destination $moduleRoot -Recurse -Force -Exclude '.git'

if (Get-Module -Name $DSCModuleName -All)
{
    Get-Module -Name $DSCModuleName -All | Remove-Module
}

Import-Module -Name $(Get-Item -Path (Join-Path $moduleRoot -ChildPath "$DSCModuleName.psd1")) -Force

if (($env:PSModulePath).Split(';') -ccontains $pwd.Path)
{
    $script:tempPath = $env:PSModulePath
    $env:PSModulePath = ($env:PSModulePath -split ';' | Where-Object {$_ -ne $pwd.path}) -join ';'
}


$executionPolicy = Get-ExecutionPolicy
if ($executionPolicy -ne 'Unrestricted')
{
    Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Force
    $rollbackExecution = $true
}


$ModuleName = "MSFT_xWebAppPool"

Import-Module (Join-Path $moduleRoot -ChildPath "DSCResources\$ModuleName\$ModuleName.psm1")

Describe "MSFT_xWebAppPool"{
    InModuleScope $ModuleName {
        Function TestTargetResourceSame
        {
            <#

            .SYNOPSIS
            Runs Test-TargetResource on $option so that when $option has the same value as the current setting, $true is returned.

            .DESCRIPTION
            Runs Test-TargetResource on $option so that when $option has the same value as the current setting, $true is returned.

            .PARAMETER option 
            should be the name of the option as defined in xWebAppPool.psm1

            .PARAMETER appcmdVal
            The location that would be used to get this value via appcmd.

            .EXAMPLE
            TestTargetResourceSame "autoStart" "add.autoStart"

            #>
            [OutputType([bool])]
            Param($option,$appcmdVal)
            $testParams =@{
                Name = "PesterAppPool"
                Ensure = "Present"
            }
            $curVal = Invoke-Expression "(([xml](& $env:SystemRoot\system32\inetsrv\appcmd.exe list apppool 'PesterAppPool' /config:*)).$($appcmdVal))"
            #Write-Host $curVal
            $testParams.Add($option,$curVal)

            #need to explicitly set it to the current value.  This "resets" all the values so the previous tests don't interfere with this one
            #Set-TargetResource @testParams | Out-Null #So we only return the value of Test-TargetResource below
            Mock Invoke-AppCmd {return $testParams}
            Test-TargetResource @testParams
            Return 
        }
```

Thanks for the help.

tysonjhayes commented:
So it'll be easier if you push the code somewhere so I can actually review it all up, that being said I had to two do things to get the code to work.

1) I moved the file MSFT_xWebAppPool.Tests.ps1 to .\xWebAdministration\Tests\Unit
2) Modified everything from line 63 (Describe "MSFT_xWebAppPool" { and below to the following.


````powershell
Describe "MSFT_xWebAppPool"{
    InModuleScope $ModuleName {
        Context 'Test-TargetResource' {
            It 'returns false if autoStart is different' {

                $testParams = @{
                    Name = "PesterAppPool"
                    Ensure = "Present"
                    AutoStart = 'False'
                }

                $failParams = $testParams
                $failParams['AutoStart'] = 'True'

                Mock Invoke-AppCmd {return $failParams}

                Test-TargetResource @testParams | Should be $false
            }
        }
    }
}
````

````powershell
[1:17:10 PM] [Merge_PSHOrg_cAppPool_to_xWebAppPool +1 ~1 -1 !] [0168] » Invoke-Pester

Describing MSFT_xWebAppPool
   Context Test-TargetResource
    [+] returns false if autoStart is different 2.36s
Tests completed in 2.36s
Passed: 1 Failed: 0 Skipped: 0 Pending: 0
````

And it works just fine. I'm unclear why its not finding the function but I'd have to see the code in more context to know why.

tysonjhayes commented:
Ok looking at the new update you gave I'm not getting an error around Mocking Invoke-AppCmd and to that end neither does AppVoyer

```powershell
  [-] Passes test when App Pool does exist and has correct Name 715ms
      Expected: {True}
      But was:  {False}
      at line: 250 in C:\projects\xwebadministration\Tests\Unit\MSFT_xWebAppPool.Tests.ps1
      250:                     Test-TargetResource @testParams | Should be $true
```

So I think you're on the right track. Things like being unable to start the AppPool will go away when you remove the code as you don't need it anyway.

zuehlaa commented:
How does this look?  Once we are good here, I'll start updating xWebsite.

tysonjhayes commented on DSCResources/MSFT_xWebAppPool/MSFT_xWebAppPool.psm1:
```diff
@@ -1,137 +1,1262 @@
 
-
-function Set-TargetResource
+function Invoke-AppCmdState
```
I'm confused why this is needed, isn't this just doing the same thing as Invoke-AppCmd?

tysonjhayes commented on DSCResources/MSFT_xWebAppPool/MSFT_xWebAppPool.psm1:
```diff
@@ -1,137 +1,1262 @@
+            Throw "Please ensure that WebAdministration module is installed."
         }
+
+        #$AppPool = & $env:SystemRoot\system32\inetsrv\appcmd.exe list apppool $Name
```
I'm not going to comment on ever line, but where you've commented out the appcmd.exe call could you please remove the line? It's not really needed at this point? :+1:

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+                Ensure = "Present"
+            }
+
+            It 'DefaultAppPool should exist' {           
```
So I have plenty of machines where I've removed the DefaultAppPool as I'm not using it, so why are we making the assumption that it should exist on every machine we're running this code on?

zuehlaa replied:
when I try to create a test app pool, appveyor gives me an access denied error.  So I went with the default app pool for the test.  It seems to exist on the appveyor build

tysonjhayes replied:
Given this is a unit test we could (and in my opinion should) mock the returns to not need to actually even check if there is any AppPools on the machine and just give the responses we're expecting.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+
+Describe "MSFT_xWebAppPool"{
+    InModuleScope $ModuleName {
+        Function TestTargetResourceSame
```
I'm not sure any of these private test functions are really needed, while reviewing this I keep wondering why they are needed. Could you help me understand the thought process behind this function?

I'll make some notes where you can make some logic changes to reduce the dependancies on these and conform to how we've been writing tests in other places (I'm not saying we're doing it right and you're wrong the great thing about open source is being able to learn from each other!)

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+            $arrVal = $appcmdVal.Split(".")             
+            $val = $global:PoolCfg.GetElementsByTagName($arrVal[$arrVal.length-2]) | Select -ExpandProperty $arrVal[$arrVal.length-1]
+            $testParams.Add($option, $val)
+            Mock Invoke-AppCmd {return $global:PoolCfg.OuterXml}
```
Both Invoke-AppCmd and Invoke-AppCmdState can be moved outside of this function.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+            }
+
+            Context 'Test-TargetResource' {
+                
```
Mock the invoke-* calls here. The mock will carry over to all calls made in the test-targetresource context this means we only create the mock once instead of during every if statement (this saves some time) in rerunning the same code.

tysonjhayes replied:
Also if you do the Invoke-AppCmd mock to
````powershell
Mock Invoke-AppCmd {return "Running"} -ParameterFilter {$Arguments.Contains('/text:state')}
`````

You can remove Invoke-AppCmdState as you can leverage [ParameterFilter](https://github.com/pester/Pester/wiki/Mock#parameterfilter) in Pester.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+                
+                It 'Passes autoStart Test when same' {
+                    
+                    TestTargetResourceSame "autoStart" "add.autoStart" | Should be $true
```
For example this could just be:
Test-TargetResource | Should Be $True

Given that Invoke-AppCmd is returning the global:PoolCfg we can design that varible to match the results we want. So this stest really is just validating that we get the match by default.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+
+                It 'Fails autoStart Test when different' {
+                    
+                    TestTargetResourceDiff "autoStart" "add.autoStart" "true" "false"| Should be $false
```
This can become:
Test-TargetResource -AutoStart $false | Should Be $false

Again because $global:PoolCfg has autostart = $true this should return false without us having to do much more.

tysonjhayes replied:
But lets say you don't want to trust that the $false is returning $false in the right place. You can actually capture the verbose output and confirm it's returning false in the right place.

````powershell
$result = Test-TargetResource -autoStart $false -Verbose -Name 'DefaultAppPool' -Verbose 4>&1
$result[0] | Should be "autoStart of AppPool DefaultAppPool does not match the desired state."
$result[1] | Should be $false
````

tysonjhayes replied:
Now it occurs to me (as I've run into this) that if we change the verbose message in the module we have to update the test as well, we'll if we move the verbose message to the LocalizedData chunk:

````powershell
AutoStartConfigFailure = autoStart of AppPool {0} does not match the desired state.
````

And change the line in test-targetresource to:
````powershell
# Check autoStart
if ($PoolConfig.add.autoStart -ne $autoStart)
{
    $DesiredConfigurationMatch = $false
    Write-Verbose $LocalizedData.AutoStartConfigFailure -f ${Name}
    break
}
````

We can change the test to:
````powershell
It 'Passes autoStart Test when same' {
    $result = Test-TargetResource -autoStart $false -Name 'DefaultAppPool' -Verbose 4>&1
    $result[0] | Should be $($LocalizedData.AutoStartConfigFailure -f 'DefaultAppPool')
    $result[1] | Should be $false
}
`````

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+                    TestTargetResourceDiff "autoStart" "add.autoStart" "true" "false"| Should be $false
+                }
+
+                It 'Passes Runtime Version Test when same' {
```
This test really isn't needed as we've actually already tested it two tests above this as it's the default parameter and already matches the $global:PoolCfg so this is covered in the first Test-TargetResource run.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+
+                It 'Fails Runtime Version Test when different' {
+                  
+                    TestTargetResourceDiff "managedRuntimeVersion" "add.managedRuntimeVersion" "v4.0" "v2.0" | Should be $false
```
This becomes
Test-TargetResource -ManagedRunTimeVersion '2.0' | Should Be $False

This passes because $global:PoolCfg is set to v4.0.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+                }
+            }
+            
+<#            Context 'Set-TargetResource' {
```
So we will want to be testing Set-TargetResource as well so we will be wanting to uncomment and run this.

zuehlaa replied:
AppVeyor gives me access denied errors when it tries to run these tests

tysonjhayes replied:
Could you give me the specific error message? I have loads of tests in othe resources that call set-targetresource without a problem, I have even written integration tests where I make changes to the system in AppVeyor, so I find it odd it's giving you access denied.

tysonjhayes commented on Tests/Unit/MSFT_xWebAppPool.Tests.ps1:
```diff
@@ -0,0 +1,862 @@
+<#            Context 'Set-TargetResource' {
+                It 'Should set autoStart Test ' {
+                    
+                    SetTargetResource "autoStart" "add.autoStart" "true" "false"| Should be $true
```
I'd recommend these tests become

````powershell
Mock Invoke-AppCmd {}
Mock Invoke-AppCmd {return $global:PoolCfg} -ParameterFilter {$Arguments.Contains('list')}

It 'Should set autoStart Test ' {
    Set-TargetResource -Name 'DefaultAppPool' -AutoStart 'False'
    Assert-MockCalled Invoke-AppCmd -ParameterFilter {$Arguments.Contains('set')} -Times 1 -Exactly
    Assert-MockCalled Invoke-AppCmd -ParameterFilter {$Arguments.Contains('/autoStart:False')} -Times 1 -Exactly
}
````

This way we can test a couple of things.
1) Invoke-AppCmd -Arguments set is only called the one time we expect it to be (given the current logic we expect set to only be called on the one difference.
2) Invoke-AppCmd is calling the specific change we expect (in this case autostart)

tysonjhayes replied:
You'll also want to write a test that checks the if the app pool doesn't exist (and needs to be created) logic.

tysonjhayes commented:
I hope my comments are constructive here without fully understanding your logic I'm hoping to explain how I've written test in other resources while trying to grock what you are attempting to do.

tysonjhayes commented:
Hey @zuehlaa - Is there anything I can help you with?

tysonjhayes commented:
I'm going to go ahead and merge this and then just rewrite put in a PR to get the tests in. Thanks!
